### DISCUSSION TITLE
Drop `PyRef` and get associated `PyCell`?

### DISCUSSION THREAD

**Discussion by cspotcode** [PyO3/pyo3#3127] [Questions] [answered]
I am writing a pyclass where most methods accept `PyRefMut` or `PyRef`.  These methods can all be called from python, they call each other in rust, and must call out to python functions, passing a reference to self, expecting self's attributes to be accessed from those functions.  So I cannot accept `&self` nor `&mut self` in these methods; I need to use `PyCell`, `PyRef`, or `PyRefMut`.  And I cannot hold a borrow of self when calling out to python functions, because they will attempt to re-borrow when they access attributes.

If each method accepts `&PyCell` then each must repeatedly borrow.  If method A calls B calls C, each is doing a borrow.  I'm fairly new to rust, but to me, this indicates I should pass a `PyRef` from A to B to C to avoid repeated borrowing and un-borrowing.

Before calling from rust to python, I need to drop my `PyRef` to avoid an "already borrowed" error.

Is there a way to drop a `PyRef` and get a ref to the `PyCell` at the same time?

```rust
// Assume self_: PyRefMut<MyClass>
let cell: &PyCell<MyClass> = self_.release_borrow_and_get_pycell();
other_python_object.call_method1(py, "foo", (cell,))?;
```

I tried to keep my question as brief as possible but I can add more detail about my use-case if it helps.

**Comment by adamreichold** [comment, marked as answer]
> Is there a way to drop a PyRef and get a ref to the PyCell at the same time?

I don't think we have such a method. The best you can do is get a `Py<T>` back from the `PyRef<T>` via a [`From` impl](https://docs.rs/pyo3/latest/pyo3/pycell/struct.PyRef.html#impl-From%3CPyRef%3C'a,+T%3E%3E-for-Py%3CT%3E).

> I tried to keep my question as brief as possible but I can add more detail about my use-case if it helps.

I am not sure that the short-lived borrows can be avoided, but I think discussing the whole problem could only help to figure out if there are any alternative solutions.

One slightly different approach might be to avoid `PyCell` completely using a `frozen` `#[pyclass]` and use a more fine-grained approach to interior mutability, i.e. all your methods take `&self` and your individual fields implement interior mutability only as required. This will not necessarily lead to longer held borrows, but it might lead to less overlap or fewer dynamically checked borrows overall depending on the shape of your data.

**Reply by cspotcode** [reply]
> get a `Py<T>` back from the `PyRef<T>` via a [`From` impl](https://docs.rs/pyo3/latest/pyo3/pycell/struct.PyRef.html#impl-From%3CPyRef%3C'a,+T%3E%3E-for-Py%3CT%3E).

I didn't know I could do this!  I'd been doing `self_.into_py(py)`.  This would let me avoid the `py` GIL token, although calling python methods requires I have it anyway.

Is there any way to convert a `PyRef` into a `Py<>` and a `Python<>` GIL token?  My goal is to drop the borrow but keep the GIL, since I want to call python methods.

> I am not sure that the short-lived borrows can be avoided

Perhaps I'm too focused on a micro optimization.  I'm considering the tradeoff between these two options:

a) Accept `self_: PyRef`/`mut self_: PyRefMut` from pyo3
**Pros**
single borrow shared across chain of rust functions.
**Cons**
perf hit for `Py::from(self_)`?  Related to GIL?  I don't know how this works.

b) accept `self_: &PyCell` from pyo3
**Pro**
Pass `self_` to `call_method1`, avoid `Py::from(self_)`?  Is there a perf concern?  I have no idea.
**Con**
each rust function makes its own short-lived borrow

> individual fields implement interior mutability only as required

Would this be using rust's `RefCell` and `RefMut`?

---

Given a `PyRef`, is there any benefit to dereferencing it once vs dereferencing implicitly on each access?

Deref once:
```rust
fn set_left(self_: PyRefMut<BasicSprite>, new_value: f32) -> PyResult<()> {
    let s = &*self_;
    let leftmost = s.hitbox.left_native();
    let x = s.position.0;
```
vs deref implicitly:
```rust
fn set_left(self_: PyRefMut<BasicSprite>, new_value: f32) -> PyResult<()> {
    let leftmost = self_.hitbox.left_native();
    let x = self_.position.0;
```

**Reply by adamreichold** [reply]
> Is there any way to convert a PyRef into a Py<> and a Python<> GIL token?

A `PyRef<'py, T>` always implies holding the GIL for `'py` and indeed you can call [`PyRef::py`](https://docs.rs/pyo3/latest/pyo3/pycell/struct.PyRef.html#method.py) to turn this information into a GIL token. So having a `PyRef` always implies having a GIL token, i.e. you can just get the token before calling into the `From` impl.

**Reply by adamreichold** [reply]
> Given a PyRef, is there any benefit to dereferencing it once vs dereferencing implicitly on each access?

As a general advice on performance, try not to rely on intuition and measure. In this particular case, I would be somewhat surprised if there was any observable difference in release builds.

**Reply by adamreichold** [reply]
> Would this be using rust's RefCell and RefMut?

This would be one alternative, you would have more options like using `Cell` for `Copy` types or `UnsafeCell` if some prohibitive overhead can only be avoid using unsafe code.

Additionally, you could use thread safe constructs like `Atomic*` and `Mutex` which have the benefit that they can be shared with other threads which do not hold the GIL. (Of course, they are usually more costly than primitives providing just interior mutability.)

**Reply by adamreichold** [reply]
> Perhaps I'm too focused on a micro optimization. I'm considering the tradeoff between these two options:

I suspect that due to the high fundamental cost of calling functions using one of CPython's calling conventions, you will have a hard time measuring the cost of something like `Py::from(self_)` which should boil down to incrementing a reference count. (Even `call_method1` itself will create a new string and a new tuple on the Python heap.)

Hence I think option a) would be preferable not so much because it is faster, but because it is simpler. That said, I would expect that repeatedly checking the same borrow flag has a similar to higher cost. As always, YMMV and you should measure if this is relevant to the runtime of your program at all.

**Reply by cspotcode** [reply]
> So having a `PyRef` always implies having a GIL token, i.e. you can just get the token before calling into the `From` impl.

Thanks, I might be doing something wrong here, I get an error on `Py::from(` saying "cannot move out of `self_` because it is borrowed"
```rust
fn set_position(mut self_: PyRefMut<'_, BasicSprite>) {
  let py = self_.py();
  let s = Py::from(self_);
  sprite_list.call_method1(py, intern!(py, "_update_position"), (&s,))?;
```
I think `from` wants to consume `self_`, yet `self_.py()` has borrowed it and the borrow is used when passing `py` into `call_method1`.

**Reply by adamreichold** [reply]
> I think from wants to consume self_, yet self_.py() has borrowed it and the borrow is used when passing py into call_method1.

I think this is a bug on our part as the returned token is limited to the borrow of the `PyRef` instead of the lifetime for which `PyRef` itself borrows the underlying cell. However, just adding a GIL token `py: Python<'_>` to your argument list should not have any computational cost.

**Reply by cspotcode** [reply]
Point taken about benchmarking versus guessing.  I think you're correct about option (a).

**Reply by adamreichold** [reply]
> I think this is a bug on our part 

c.f. #3131

**Reply by cspotcode** [reply]
> I think this is a bug on our part as the returned token is limited to the borrow of the PyRef instead of the lifetime for which PyRef itself borrows the underlying cell. However, just adding a GIL token py: Python<'_> to your argument list should not have any computational cost.

Ok, I'm happy to leave it at that: we will accept and pass around a `py: Python<'_>` argument.  It would be ergonomically nice to avoid passing around `py` since the `self_: PyRef` already implies having the GIL, but it's not a big deal.
